### PR TITLE
Move object history in difficulty calculation to the object itself

### PR DESCRIPTION
**Pull request by apollo-dw**
first PR for a skills refactor, second attempt of https://github.com/ppy/osu/pull/18057, and based on this discussion: https://github.com/ppy/osu/discussions/16122#discussioncomment-1953131.

Concept
---
We want to calculate the difficulty of any arbitrary object without needing to process all the previous ones. This is not possible in the current system, since object history is stored in the difficulty calculation when processing objects successively. This PR is a step forward to achieving this by moving history to the object itself, allowing it to fetch a previous object (or even potentially a next object).

Storing a reference to the overall list in the object will be iterated upon and improved later since it's by no means a perfect solution: this is in the interest of moving things forward

Next step
---
Next up would be splitting the individual object difficulty calculation from the total difficulty values aggregation, which are currently combined in `Skill`s. This has been done for the osu! ruleset and can be previewed on [this WIP branch](https://github.com/apollo-dw/osu/tree/wip) - the jist is that those object difficulty calculations were moved to "Evaluators" that can be called from the `Skill` to produce the list of values.

### CODE REVIEW

File: osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs [hunk 1]
```diff
@@ -46,13 +46,13 @@ public TaikoDifficultyCalculator(IRulesetInfo ruleset, IWorkingBeatmap beatmap)
...
         {
-            List<TaikoDifficultyHitObject> taikoDifficultyHitObjects = new List<TaikoDifficultyHitObject>();
+            List<DifficultyHitObject> taikoDifficultyHitObjects = new List<DifficultyHitObject>();
 
             for (int i = 2; i < beatmap.HitObjects.Count; i++)
             {
                 taikoDifficultyHitObjects.Add(
                     new TaikoDifficultyHitObject(
-                        beatmap.HitObjects[i], beatmap.HitObjects[i - 1], beatmap.HitObjects[i - 2], clockRate, i
+                        beatmap.HitObjects[i], beatmap.HitObjects[i - 1], beatmap.HitObjects[i - 2], clockRate, taikoDifficultyHitObjects, i - 2
```
Comment (bdach) [inline osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs:55]: I'll just note that the change in the last argument very luckily does not seem like it will influence results, because the only usage of the old `ObjectIndex` was to calculate a difference between two object indices in `Rhythm`, and so `b.ObjectIndex - a.ObjectIndex == (b.Position + 2) - (a.Position + 2)`.

File: osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/TaikoDifficultyHitObject.cs [hunk 2]
```diff
@@ -42,16 +38,15 @@ public class TaikoDifficultyHitObject : DifficultyHitObject
         /// <param name="lastObject">The gameplay <see cref="HitObject"/> preceding <paramref name="hitObject"/>.</param>
         /// <param name="lastLastObject">The gameplay <see cref="HitObject"/> preceding <paramref name="lastObject"/>.</param>
         /// <param name="clockRate">The rate of the gameplay clock. Modified by speed-changing mods.</param>
-        /// <param name="objectIndex">The index of the object in the beatmap.</param>
-        public TaikoDifficultyHitObject(HitObject hitObject, HitObject lastObject, HitObject lastLastObject, double clockRate, int objectIndex)
-            : base(hitObject, lastObject, clockRate)
+        /// <param name="objects">The list of <see cref="DifficultyHitObject"/>s in the current beatmap.</param>
+        /// <param name="position">The index of the object in the beatmap.</param>
```
Comment (bdach) [inline osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/TaikoDifficultyHitObject.cs:42]: This xmldoc isn't really true anymore, though. `position` is off by 2 compared to the real index of the object in the beatmap at best.

File: osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs [hunk 3]
```diff
@@ -12,7 +13,7 @@ namespace osu.Game.Rulesets.Osu.Difficulty.Preprocessing
 {
     public class OsuDifficultyHitObject : DifficultyHitObject
     {
-        private const int normalised_radius = 50; // Change radius to 50 to make 100 the diameter. Easier for mental maths.
+        private const int normalised_radius = 50;
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:16]: Not sure why comment was trimmed here?

**Review by bdach — COMMENTED** [review]
I could be okay with this start, albeit it does appear a bit undercooked in places.

File: osu.Game/Rulesets/Difficulty/Preprocessing/DifficultyHitObject.cs [hunk 4]
```diff
@@ -41,13 +49,21 @@ public class DifficultyHitObject
         /// <param name="hitObject">The <see cref="HitObject"/> which this <see cref="DifficultyHitObject"/> wraps.</param>
         /// <param name="lastObject">The last <see cref="HitObject"/> which occurs before <paramref name="hitObject"/> in the beatmap.</param>
         /// <param name="clockRate">The rate at which the gameplay clock is run at.</param>
-        public DifficultyHitObject(HitObject hitObject, HitObject lastObject, double clockRate)
+        /// <param name="objects">The list of <see cref="DifficultyHitObject"/>s in the current beatmap.</param>
+        /// <param name="position">The position of this <see cref="DifficultyHitObject"/> in the <see cref="difficultyHitObjects"/> list.</param>
+        public DifficultyHitObject(HitObject hitObject, HitObject lastObject, double clockRate, List<DifficultyHitObject> objects, int position)
```
[the remark this answers is not in the document]
Comment (apollo-dw) [reply osu.Game/Rulesets/Difficulty/Preprocessing/DifficultyHitObject.cs:54]: With regards to `lastObject`: I did indeed try this however it ended up being impossible because of the way some ruleset's difficulty calculators skip the first 1 or 2 objects: the very first `DifficultyHitObject` can have a `lastObject` whilst being the first object in the list, so `objects[position - 1` means.... yeah. and I'm not looking to change performances values with the changes here.

As for `hitObject`, yeah I don't see why not

[the remark this answers is not in the document]
Comment (apollo-dw) [reply osu.Game/Rulesets/Difficulty/Preprocessing/DifficultyHitObject.cs:54]: Actually, I do see why not: the objects list is a list of `DifficultyHitObject`s, and there's no way to cast them to a `HitObject` as far as I can tell. You can only retrieve their `BaseObject` which is passed through the constructor here. Seems like we need both here

[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Rulesets/Difficulty/Preprocessing/DifficultyHitObject.cs:54]: I see. Fair enough then I guess, although it's a touch hideous. Do you foresee this going away in some capacity in the future? Maybe even via removing difficulty hitobjects altogether? (bit radical but maybe?)

[the remark this answers is not in the document]
Comment (apollo-dw) [reply osu.Game/Rulesets/Difficulty/Preprocessing/DifficultyHitObject.cs:54]: Yeah I could also do with this not overstaying it's welcome. Probably not by removing them since I find them to be pretty valuable, so I'll look into options at a later point (being able to upcast would solve this entirely I think)

[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game/Rulesets/Difficulty/Preprocessing/DifficultyHitObject.cs:54]: This is somewhat similar to what I've said in https://github.com/ppy/osu/pull/18057#discussion_r863302374, and figured it can't be removed for the time being due to the ruleset-specific way of not 1:1 mapping hitobjects to `DifficultyHitObject`s.

File: osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs [hunk 5]
```diff
@@ -74,10 +75,10 @@ public class OsuDifficultyHitObject : DifficultyHitObject
         private readonly OsuHitObject lastLastObject;
         private readonly OsuHitObject lastObject;
 
-        public OsuDifficultyHitObject(HitObject hitObject, HitObject lastLastObject, HitObject lastObject, double clockRate)
-            : base(hitObject, lastObject, clockRate)
+        public OsuDifficultyHitObject(HitObject hitObject, HitObject lastObject, double clockRate, List<DifficultyHitObject> objects, int position)
+            : base(hitObject, lastObject, clockRate, objects, position)
         {
-            this.lastLastObject = (OsuHitObject)lastLastObject;
+            lastLastObject = position > 1 ? (OsuHitObject)Previous(1).BaseObject : null;
```
Comment (frenzibyte) [inline osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:81]: Could simplify this to `lastLastObject = (OsuHitObject)Previous(1)?.BaseObject`, if `Previous` allows returning null when index is out of range (see other comment).

**Review by frenzibyte — COMMENTED** [review]
While I am fine with the current way of retrieving previous/next objects in this PR, I'm just going to note that for future consideration this could be further improved w.r.t. code quality/readability using this set instead:
```csharp
        // use instead of Previous(0)
        public DifficultyHitObject Previous => difficultyHitObjects.ElementAtOrDefault(Position - 1);
 
        public DifficultyHitObject PreviousAt(int index) => difficultyHitObjects.ElementAtOrDefault(Position - (index + 1));

        // use instead of for (int i = 0; i < Math.Min(Positon, count); i++)
        public IEnumerable<DifficultyHitObject> PreviousUntil(int count) => difficultyHitObjects.Reverse()
                                                                                                .Skip(difficultyHitObjects.Count - Position)
                                                                                                .Take(count);
```

<details>
<summary>diff</summary>

```diff
diff --git a/osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs b/osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs
index 626f1bebbd..a4723fbad0 100644
--- a/osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs
+++ b/osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs
@@ -85,8 +85,8 @@ protected override double StrainValueOf(DifficultyHitObject current)
         }
 
         protected override double CalculateInitialStrain(double offset, DifficultyHitObject current)
-            => applyDecay(individualStrain, offset - current.Previous(0).StartTime, individual_decay_base)
-               + applyDecay(overallStrain, offset - current.Previous(0).StartTime, overall_decay_base);
+            => applyDecay(individualStrain, offset - current.Previous.StartTime, individual_decay_base)
+               + applyDecay(overallStrain, offset - current.Previous.StartTime, overall_decay_base);
 
         private double applyDecay(double value, double deltaTime, double decayBase)
             => value * Math.Pow(decayBase, deltaTime / 1000);
diff --git a/osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs b/osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
index 591a6e4aa2..47b5a1d1cd 100644
--- a/osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
+++ b/osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
@@ -78,7 +78,7 @@ public class OsuDifficultyHitObject : DifficultyHitObject
         public OsuDifficultyHitObject(HitObject hitObject, HitObject lastObject, double clockRate, List<DifficultyHitObject> objects, int position)
             : base(hitObject, lastObject, clockRate, objects, position)
         {
-            lastLastObject = position > 1 ? (OsuHitObject)Previous(1).BaseObject : null;
+            lastLastObject = (OsuHitObject)PreviousAt(1)?.BaseObject;
             this.lastObject = (OsuHitObject)lastObject;
 
             // Capped to 25ms to prevent difficulty calculation breaking from simultaneous objects.
diff --git a/osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs b/osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs
index e3ec997740..8e559b4046 100644
--- a/osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs
+++ b/osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs
@@ -34,12 +34,12 @@ public Aim(Mod[] mods, bool withSliders)
 
         private double strainValueOf(DifficultyHitObject current)
         {
-            if (current.BaseObject is Spinner || current.Position <= 1 || current.Previous(0).BaseObject is Spinner)
+            if (current.BaseObject is Spinner || current.Position <= 1 || current.Previous.BaseObject is Spinner)
                 return 0;
 
             var osuCurrObj = (OsuDifficultyHitObject)current;
-            var osuLastObj = (OsuDifficultyHitObject)current.Previous(0);
-            var osuLastLastObj = (OsuDifficultyHitObject)current.Previous(1);
+            var osuLastObj = (OsuDifficultyHitObject)current.PreviousAt(0)!;
+            var osuLastLastObj = (OsuDifficultyHitObject)current.PreviousAt(1)!;
 
             // Calculate the velocity to the current hitobject, which starts with a base distance / time assuming the last object is a hitcircle.
             double currVelocity = osuCurrObj.LazyJumpDistance / osuCurrObj.StrainTime;
@@ -150,7 +150,7 @@ private double strainValueOf(DifficultyHitObject current)
 
         private double strainDecay(double ms) => Math.Pow(strainDecayBase, ms / 1000);
 
-        protected override double CalculateInitialStrain(double time, DifficultyHitObject current) => currentStrain * strainDecay(time - current.Previous(0).StartTime);
+        protected override double CalculateInitialStrain(double time, DifficultyHitObject current) => currentStrain * strainDecay(time - current.Previous.StartTime);
 
         protected override double StrainValueAt(DifficultyHitObject current)
         {
diff --git a/osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs b/osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs
index 75796a5993..9ab72a4e8f 100644
--- a/osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs
+++ b/osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs
@@ -50,9 +50,8 @@ private double strainValueOf(DifficultyHitObject current)
             OsuDifficultyHitObject lastObj = osuCurrent;
 
             // This is iterating backwards in time from the current object.
-            for (int i = 0; i < Math.Min(current.Position, 10); i++)
+            foreach (var currentObj in current.PreviousUntil(10).Cast<OsuDifficultyHitObject>())
             {
-                var currentObj = (OsuDifficultyHitObject)current.Previous(i);
                 var currentHitObject = (OsuHitObject)(currentObj.BaseObject);
 
                 if (!(currentObj.BaseObject is Spinner))
@@ -62,7 +61,7 @@ private double strainValueOf(DifficultyHitObject current)
                     cumulativeStrainTime += lastObj.StrainTime;
 
                     // We want to nerf objects that can be easily seen within the Flashlight circle radius.
-                    if (i == 0)
+                    if (currentObj.Position == lastObj.Position)
                         smallDistNerf = Math.Min(1.0, jumpDistance / 75.0);
 
                     // We also want to nerf stacks so that only the first object of the stack is accounted for.
@@ -88,7 +87,7 @@ private double strainValueOf(DifficultyHitObject current)
 
         private double strainDecay(double ms) => Math.Pow(strainDecayBase, ms / 1000);
 
-        protected override double CalculateInitialStrain(double time, DifficultyHitObject current) => currentStrain * strainDecay(time - current.Previous(0).StartTime);
+        protected override double CalculateInitialStrain(double time, DifficultyHitObject current) => currentStrain * strainDecay(time - current.Previous.StartTime);
 
         protected override double StrainValueAt(DifficultyHitObject current)
         {
diff --git a/osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs b/osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs
index f8905664f4..3a5d3131fe 100644
--- a/osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs
+++ b/osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs
@@ -57,14 +57,14 @@ private double calculateRhythmBonus(DifficultyHitObject current)
 
             int rhythmStart = 0;
 
-            while (rhythmStart < historicalNoteCount - 2 && current.StartTime - current.Previous(rhythmStart).StartTime < history_time_max)
+            while (rhythmStart < historicalNoteCount - 2 && current.StartTime - current.PreviousAt(rhythmStart).StartTime < history_time_max)
                 rhythmStart++;
 
             for (int i = rhythmStart; i > 0; i--)
             {
-                OsuDifficultyHitObject currObj = (OsuDifficultyHitObject)current.Previous(i - 1);
-                OsuDifficultyHitObject prevObj = (OsuDifficultyHitObject)current.Previous(i);
-                OsuDifficultyHitObject lastObj = (OsuDifficultyHitObject)current.Previous(i + 1);
+                OsuDifficultyHitObject currObj = (OsuDifficultyHitObject)current.PreviousAt(i - 1);
+                OsuDifficultyHitObject prevObj = (OsuDifficultyHitObject)current.PreviousAt(i);
+                OsuDifficultyHitObject lastObj = (OsuDifficultyHitObject)current.PreviousAt(i + 1);
 
                 double currHistoricalDecay = (history_time_max - (current.StartTime - currObj.StartTime)) / history_time_max; // scales note 0 to 1 from history to now
 
@@ -90,10 +90,10 @@ private double calculateRhythmBonus(DifficultyHitObject current)
                     }
                     else
                     {
-                        if (current.Previous(i - 1).BaseObject is Slider) // bpm change is into slider, this is easy acc window
+                        if (current.PreviousAt(i - 1).BaseObject is Slider) // bpm change is into slider, this is easy acc window
                             effectiveRatio *= 0.125;
 
-                        if (current.Previous(i).BaseObject is Slider) // bpm change was from a slider, this is easier typically than circle -> circle
+                        if (current.PreviousAt(i).BaseObject is Slider) // bpm change was from a slider, this is easier typically than circle -> circle
                             effectiveRatio *= 0.25;
 
                         if (previousIslandSize == islandSize) // repeated island size (ex: triplet -> triplet)
@@ -136,7 +136,7 @@ private double strainValueOf(DifficultyHitObject current)
 
             // derive strainTime for calculation
             var osuCurrObj = (OsuDifficultyHitObject)current;
-            var osuPrevObj = current.Position > 0 ? (OsuDifficultyHitObject)current.Previous(0) : null;
+            var osuPrevObj = current.Position > 0 ? (OsuDifficultyHitObject)current.Previous : null;
 
             double strainTime = osuCurrObj.StrainTime;
             double greatWindowFull = greatWindow * 2;
@@ -164,7 +164,7 @@ private double strainValueOf(DifficultyHitObject current)
 
         private double strainDecay(double ms) => Math.Pow(strainDecayBase, ms / 1000);
 
-        protected override double CalculateInitialStrain(double time, DifficultyHitObject current) => (currentStrain * currentRhythm) * strainDecay(time - current.Previous(0).StartTime);
+        protected override double CalculateInitialStrain(double time, DifficultyHitObject current) => (currentStrain * currentRhythm) * strainDecay(time - current.Previous.StartTime);
 
         protected override double StrainValueAt(DifficultyHitObject current)
         {
diff --git a/osu.Game/Rulesets/Difficulty/Skills/StrainDecaySkill.cs b/osu.Game/Rulesets/Difficulty/Skills/StrainDecaySkill.cs
index 3f3d57938b..bebf08c5a0 100644
--- a/osu.Game/Rulesets/Difficulty/Skills/StrainDecaySkill.cs
+++ b/osu.Game/Rulesets/Difficulty/Skills/StrainDecaySkill.cs
@@ -34,7 +34,7 @@ protected StrainDecaySkill(Mod[] mods)
         {
         }
 
-        protected override double CalculateInitialStrain(double time, DifficultyHitObject current) => CurrentStrain * strainDecay(time - current.Previous(0).StartTime);
+        protected override double CalculateInitialStrain(double time, DifficultyHitObject current) => CurrentStrain * strainDecay(time - current.Previous.StartTime);
 
         protected override double StrainValueAt(DifficultyHitObject current)
         {
```

</details>

You can also expose the objects list and move the methods to an extensions class with generics to avoid having to cast to `OsuDifficultyHitObject` (or other ruleset types) in usages, by making it dependant on the current object's type instead, i.e.:
```csharp
        public static IEnumerable<T> PreviousUntil<T>(this T current, int count)
            where T : DifficultyHitObject
        {
            return current.List.Reverse()
                          .Skip(current.List.Count - current.Position)
                          .Take(count)
                          .Cast<T>();
        }
```

```csharp
DifficultyHitObject current;
OsuDifficultyHitObject osuCurrent;
current.PreviousUntil(10); // returns IEnumerable<DifficutlyHitObject>
osuCurrent.PreviousUntil(10); // returns IEnumerable<OsuDifficultyHitObject>

File: osu.Game/Rulesets/Difficulty/Preprocessing/DifficultyHitObject.cs [hunk 6]
```diff
@@ -41,13 +49,21 @@ public class DifficultyHitObject
...
         {
+            difficultyHitObjects = objects;
+            Position = position;
```
[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Rulesets/Difficulty/Preprocessing/DifficultyHitObject.cs]: I can't really find myself agreeing with this change. Relying on the object count when constructing difficulty hitobject locks the users into having to construct the object sequentially, but the API does not communicate clearly that this is the case. I preferred explicit index passing.

[the remark this answers is not in the document]
Comment (apollo-dw) [reply osu.Game/Rulesets/Difficulty/Preprocessing/DifficultyHitObject.cs]: I think I'd prefer making it clear that the method should generate difficulty hitobjects sequentially, unless I'm missing some specific use cases which would be entirely reasonable - open to more opinions on this

[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game/Rulesets/Difficulty/Preprocessing/DifficultyHitObject.cs]: I get where @bdach is coming from, short of that `DifficultyHitObject` is now computing its position by guessing that it will only ever be added to the end of the list, and it's not documented anywhere that that can be the case.

I would be fine with bringing back the constructor parameter but still pass `list.Count` in usages rather than trying to compute it from the for-loop index.

I'm also gonna note that you can alternatively avoid the necessity of a `Position` by computing the index of the object directly instead (`list.IndexOf(this)` during `Previous`/`Next` calls), but that might be unnecessarily expensive, not sure.

[the remark this answers is not in the document]
Comment (apollo-dw) [reply osu.Game/Rulesets/Difficulty/Preprocessing/DifficultyHitObject.cs:58]: Done as suggested - will keep the `Position` property around though since it has use across various parts of difficulty calculation

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Comment by smoogipoo** [comment]
So in order to fix the above issue, I tried to refactor this following a different path of tying `HitObject`s 1-1 to `DifficultyHitObject`s. It's not a simple change though and generates more differences locally than this PR, so I'm halting work on that for the time being.

I think it's something that we may want to do though, probably fit for future work... I envision DHOs as just storing data, which most of them do except `OsuDifficultyHitObject` & `TaikoDifficultyHitObject` that also do some computation alongside requiring the `last`/`lastLast` objects.

The issue pointed above needs to be fixed, though.

File: osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs [hunk 7]
```diff
@@ -87,16 +87,14 @@ protected override DifficultyAttributes CreateDifficultyAttributes(IBeatmap beat
...
             // If the map has less than two OsuHitObjects, the enumerator will not return anything.
             for (int i = 1; i < beatmap.HitObjects.Count; i++)
-            {
-                var lastLast = i > 1 ? beatmap.HitObjects[i - 2] : null;
-                var last = beatmap.HitObjects[i - 1];
-                var current = beatmap.HitObjects[i];
+                objects.Add(new OsuDifficultyHitObject(beatmap.HitObjects[i], beatmap.HitObjects[i - 1], clockRate, objects, objects.Count));
 
-                yield return new OsuDifficultyHitObject(current, lastLast, last, clockRate);
-            }
```
[the remark this answers is not in the document]
Comment (apollo-dw) [reply osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs:99]: Super sneaky... reverted behaviour here to be similar to master which appears to have fixed the map above. Curious that tests passed initially with this broken, might be something specific with this map worth taking into account in the test map?

[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs]: Probably, but I'm not really sold on the test map as a means to an end for now. It's the reason why I do this sort of testing myself because there could be edge cases in some maps out there that would otherwise be missed.

**Review by smoogipoo — APPROVED** [review]
Looks fine for the time being. I've tested across all ranked beatmaps as of 2022-06, including converts, across all rulesets, to confirm that there's 0 changes resulting from this PR.

I think it can probably be improved upon further (see my earlier comments in the thread) but needs a bit more thinking.
